subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedTaskExecutionIntegrationTest.groovy

```diff
@@ -485,4 +485,34 @@ class CachedTaskExecutionIntegrationTest extends AbstractIntegrationSpec impleme
         then:
         result.assertTaskSkipped(":cacheable")
     }
+
+    def "downstream task stays cached when upstream task is loaded from cache"() {
```

Comment:
From the test it is not clear what the downstream and upstream tasks are (I guess `test` and `compileTestJava`). Did this fail at some point during the changes?
It would be great if we could not use task types which are only part of some plugins in `core` tests. Having custom task types in this test probably also makes it clearer what the test aims to do.